### PR TITLE
[Post] Accessing hardware devices on the web

### LINKED ISSUE #4489: content: hardware devices 101 [2021-01-29]
* [draft](https://docs.google.com/document/d/1nZSK6erFtJzKY5-PiBzB-kk7b8h2d60GezgOTBKi5W0/edit)
* [thread](https://groups.google.com/a/google.com/g/webdev-content/c/0eRCFezzCC8)

### PR DESCRIPTION
**Pull request by beaufortfrancois**
This article is 101 article that acts as an introduction to web.dev/devices. It aims to help web developers to communicate with a hardware device of their choice by picking the appropriate Web API.

Fixes #4489

Live preview:
- https://deploy-preview-4472--web-dev-staging.netlify.app/devices-introduction/
- https://deploy-preview-4472--web-dev-staging.netlify.app/devices/

@reillyeon Can you have a look?

### CODE REVIEW

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:6]:
```suggestion
  This article helps Web developers pick the right hardware API based on a given device.
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:31]:
```suggestion
If so, use [`MediaDevices.getUserMedia()`] to get a live video stream from this
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:33]:
```suggestion
device and learn about [capturing video]. You can also [control the camera's pan,
tilt, and zoom], and other camera settings such as [brightness and contrast], and even
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:21]:
```suggestion
with a hardware device of your choice (e.g., a webcam, a microphone, a mouse, a
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:36]:
```suggestion
## 2. Do you want to access audio streams from this device? {: #audio }
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:38]:
```suggestion
If so, use [`MediaDevices.getUserMedia()`] to get a live audio stream from this
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:40]:
```suggestion
effects to audio, create audio visualizations, or apply spatial effects (such as
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:46]:
```suggestion
If so, try listening for [Keyboard] and [Pointer] events. If this device is a game
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:48]:
```suggestion
controller, use the [Gamepad] API to know which buttons are being pressed and which axes moved.
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:66]: I love it! I wonder if it will survive the web.dev team's review.

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:72]:
```suggestion
behavior is [well-documented]. If not, at this point, you either have to find
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:74]:
```suggestion
some hardware documentation for this device or reverse-engineer it. See the [Reverse-Engineering a Bluetooth Lightbulb] article from Uri Shaked for an example on how to do that.
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:134]:
```suggestion
Thank you to [Reilly Grant](https://github.com/reillyeon) for the review of this article.
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:139]:
```suggestion
Photo by [Darya Tryfanava](https://unsplash.com/@darya_tryfanava) on [Unsplash](https://unsplash.com/).
```

Add the deeplink to the photo, for whatever reason the Unsplash widget only links to the photographer.

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:142]:
```suggestion
[`MediaDevices.getUserMedia()`]: https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/getUserMedia
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:156]:
```suggestion
[Reverse-Engineering a Bluetooth Lightbulb]: https://urish.medium.com/reverse-engineering-a-bluetooth-lightbulb-56580fcb7546
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:145]:
```suggestion
[brightness and contrast]: https://developers.google.com/web/updates/2016/12/imagecapture
```

File: src/site/content/en/blog/access-devices/index.md
Comment (tomayac) [inline src/site/content/en/blog/access-devices/index.md:144]:
```suggestion
[control the camera's pan, tilt, and zoom]: /camera-pan-tilt-zoom/
```

**Review by tomayac — CHANGES_REQUESTED** [review]
LGTM after the requested changes.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:31]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:33]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:38]: Done

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:36]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:40]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:46]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:48]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:66]: :D

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:72]: Done

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:74]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:134]: I've added you.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:139]: Done.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:142]: Done

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:144]: Done

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:145]: Done

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:156]: Done

**Review by beaufortfrancois — COMMENTED** [review]
Thanks for reviewing @tomayac!

**Review by kaycebasques — CHANGES_REQUESTED** [review]
As mentioned in the [thread](https://groups.google.com/a/google.com/g/webdev-content/c/qJptZsS4Sco) I like the idea but I would like to get a TW (either me or Joe) to work with you on investigating whether we can streamline the presentation of the information.

**Comment by beaufortfrancois** [comment]
> As mentioned in the [thread](https://groups.google.com/a/google.com/g/webdev-content/c/qJptZsS4Sco) I like the idea but I would like to get a TW (either me or Joe) to work with you on investigating whether we can streamline the presentation of the information.

See my answer: https://groups.google.com/a/google.com/g/webdev-content/c/qJptZsS4Sco/m/j4QxfGNIDgAJ

File: src/site/content/en/blog/access-devices/index.md
Comment (reillyeon) [inline src/site/content/en/blog/access-devices/index.md:54]: A few other types of devices to include here are storage devices (File System Access API), printers (window.print()), network adapters (fetch()) and authenticators (WebAuthn).

File: src/site/content/en/blog/access-devices/index.md
Comment (reillyeon) [reply src/site/content/en/blog/access-devices/index.md:66]: Where I am there are so many devices in the local environment it is a little hard to interpret the results. I would point developers at a larger article that explains how to use chrome://bluetooth-internals and/or the nRF Connect app for Bluetooth discovery and reverse engineering.

File: src/site/content/en/blog/access-devices/index.md
Comment (reillyeon) [inline src/site/content/en/blog/access-devices/index.md:88]: As cute as this is a device appearing in the chooser doesn't always mean that it is accessible. Tools like chrome://bluetooth-internals and https://webhid-collections.glitch.me/ can be really helpful for exploring what is actually possible a device.

**Comment by jpmedley** [comment]
Here's my first two recommendations about this article. 

* Introduce the article by explaining how the API revealed capabilities you didn't know a device had. 
* Change language so that it's clear you're showing how to discover features.

**Comment by reillyeon** [comment]
The intention of this article is not to show how to discover device features. The target audience is a developer who has a device and wants to know how to take advantage of that device's capabilities from a web application. The goal is to guide the developer through the available options. That starts with the high-level APIs mentioned at the beginning of the article and ends with the low-level APIs and links to resources for learning more.

One of the problems that this article is trying to solve is developers trying the low-level solutions first (i.e. jumping straight to WebUSB) rather than learning about the high-level options which are easier to use and provide better UX.

It is also important for developers who need to take the low-level approach to appreciate that there is rarely a one-side-fits-all solution. This article is great because it highlights the need for good device documentation and/or reverse engineering in order to build a solution using the low-level APIs we provide.

**Review by kaycebasques — COMMENTED** [review]
@beaufortfrancois I've got some questions to help us decide how to proceed. At the end we (you and me togethery) will either decide to 1) defer to you and let you determine what's best for the content or 2) keep hashing out the content together after determining the content isn't yet meeting your goals.

## Audience

Who is the audience? Someone who is 1) already quite knowledgeable about devices? Or 2) web developers who may not know much about devices?

If it's #2, then I am a pretty good representation of your target market (I actually may know a bit more than your typical web developer because I worked at an internet of things startup). And because I'm having trouble with the content, that suggests others in my shoes might also struggle. If it's #1 then Joe and I probably just didn't have enough domain knowledge to review this properly (which would explain the back-and-forth).

## Role of this content

Your goal here is for this to be the entrypoint into the devices collection? Is that correct? Or are you going to position this as a secondary guide?

If it's supposed to be the entrypoint, then my earlier question about audience is important here. Because if this is supposed to be the intro for web developers who don't know much about devices, I think we want to think hard about our "mental model" of organizing all the device APIs.

## Goal of the page

I'm still a bit unclear about the goal here. Let me try to provide lots of explanation based on quotes that you and Reilly have provided:

> My goal is to help you, web developer, pick the appropriate API to communicate
with a hardware device of your choice (e.g., a webcam, a microphone, a mouse, a
keyboard, a stylus, an NFC tag, a Bluetooth device, or an exotic hardware
device).

This implies that you have choices. But for a lot of these APIs there's only one choice:

image

image

image

image

image

Also, the way that you have organized (numbering the sections from 1 to 8) these implies that you recommend people to use NFC first, and then go to Bluetooth if NFC doesn't meet their needs, then HID, then USB, then serial. Is that intended?

I think this comment from @reillyeon might hold the key:

> One of the problems that this article is trying to solve is developers trying the low-level solutions first (i.e. jumping straight to WebUSB) rather than learning about the high-level options which are easier to use and provide better UX.

If this is the case, then I think my big confusion is that this article isn't focused on my use cases. It's focused on the technologies. And as a web developer who doesn't know much about each of the APIs, I have no idea when I would want to use one versus the other.

Also, as I mentioned in a comment, the fact that these sections are numbered sequentially is confusing to me. Numbered sections implies that I start with 1, then go to 2, and so on. If my use case is accessing video, why do I have to start with #1?

The first three sections are examples of content that **are** focused on use cases that I think our audience will search for. Audio, video, input. But then we just jump into 3 APIs without context about what user-focused tasks those APIs relate to. E.g. can I use NFC for audio? Can I use USB for video?

If the goal is to steer people from high-level APIs to low-level APIs, I think the best way to accomplish that is to think about the use cases in the general terms that we think web developers will actually use.

For example, it seems like this might be one of the high-level to low-level paths you're thinking of:

Want to handle input? Then check out these APIs (in this order):

* Keyboard
* Pointer
* Gamepad
* HID

And then from here, if none of these solve your problem, the decision of what low-level API you should use seems to be determined by the nature of your physical connection to the device:

* NFC for very close wireless connections
* Bluetooth for close wireless
* USB for wired
* Serial for wired (what's the difference between this and USB?)

Or maybe some of these lower-level APIs aren't appropriate for input? In which case we have further helped the reader (by removing them from the list of lower-level APIs to consider and helping them focus only on relevant ones).

And then we would do the same for accessing audio, accessing video, and any other user-focused tasks you can think of. 

To put the problem another way, the content flow seems to break down for me when we get to the NFC, Bluetooth, HID, USB, and serial sections. What **user-focused** tasks do those APIs help me solve? Are these the low-level APIs you've been mentioning and they all can accomplish the same user-focused tasks? In that case, what is the deciding factor on when to use one versus the other?

LMK if my feedback is way off target (and therefore I'll take Joe and me off reviewing this and defer to you) or if you want to keep iterating with me/Joe based on the feedback!

**Comment by reillyeon** [comment]
The audience is definitely #2, a developer who has a device-related goal (e.g. capturing an image with a camera) and needs to find the right API for the task. Using that goal as an example, depending on the type of camera the answer could be to use the Image Capture API to capture a frame from the video stream, or it could be using WebUSB to speak PTP (picture transfer protocol) directly. The challenge I see for an article like this is trying to cover as many potential goals as possible without listing literally every single web API in existence.

Your interpretation of my comment about device features is correct. To use a camera again as an example, I expect the developer to know generally that a camera is capable of capturing images and sending them to the computer but not to understand the nuance between the USB UVC and PTP device classes which affects their options for how to accomplish that task.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:11]: Me too ;)

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:27]: I like this intro 👍

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:29]: True. We'll think of other ways to break this down.

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:54]: I didn't know about network adapters ;)

File: src/site/content/en/blog/access-devices/index.md
Comment (beaufortfrancois) [reply src/site/content/en/blog/access-devices/index.md:136]: Joe didn't yet.

**Review by beaufortfrancois — COMMENTED** [review]
Thank you really much for your questions and feedback @kaycebasques 
Based on those, I think I need to go back to the blackboard to figure out how to position this article in web.dev/devices and which APIs we want to list there.
@reillyeon What do you think?

**Comment by reillyeon** [comment]
I think that makes sense and I'll retract my earlier comment about this article listing every web API. I think we do need to have a "map" of web APIs, at least in the device space, and this could be the opportunity.

File: src/site/content/en/blog/access-devices/index.md
Comment (robdodson) [reply src/site/content/en/blog/access-devices/index.md:66]: In general I prefer for these to be part of an iframe embed (glitch or codepen).

**Comment by beaufortfrancois** [comment]
@kaycebasques Please review.

File: src/site/content/en/blog/access-devices/index.md
Comment (reillyeon) [inline src/site/content/en/blog/access-devices/index.md:146]: The intention of this note is to let developers know that they may want to check out the WebHID or Web Serial APIs, even for Bluetooth devices. Maybe reversing this sentence (i.e. "Note that Bluetooth devices may also speak the HID or serial protocols.") would make that clearer.

**Review by reillyeon — CHANGES_REQUESTED** [review]

**Review by kaycebasques — APPROVED** [review]
LGTM! @reillyeon suggested a change.

**Comment by beaufortfrancois** [comment]
I've addressed latest feedback from @reillyeon and @kaycebasques. Thanks again!

**Pull request merged**
